### PR TITLE
Implement Adabelief

### LINKED ISSUE #6: Implement scaling by AdaBelief
See paper:

https://arxiv.org/abs/2010.07468

This should be a trivial modification of the existing `scale_by_adam` transform.

### PR DESCRIPTION
**Pull request by joaogui1**
Closes #6 

1. Is the implementation ok? (will fix the docstring together with tests)
2. How exactly is testing done? Just add it to the `@parameterized.named_parameters` arguments with default learning rate?

### CODE REVIEW

File: optax/_src/transform.py [hunk 1]
```diff
@@ -364,6 +366,44 @@ def update_fn(updates, state, params=None):
...
+
+  def init_fn(params):
+    mu = jax.tree_map(jnp.zeros_like, params)  # First moment
+    s = jax.tree_map(jnp.zeros_like, params)  # Second Central moment
+    return ScaleByAdamState(count=jnp.zeros([], jnp.int32), mu=mu, nu=s)
+
+  def update_fn(updates, state, params=None):
+    del params
+    mu = _update_moment(updates, state.mu, b1, 1)
+    nu = _update_moment(updates - state.mu, state.nu, b2, 2)
```
Comment (mtthss) [inline optax/_src/transform.py:396]: `updates` and `state.mu` might be arbitrary data structure,
therefore you cannot subtract them directly

Please use tree_multimap as follows:

```
prediction_error = jax.tree_multimap(lambda g, m: g-m, updates, mu)
nu = _update_moment(prediction_error, state.nu, b2, 2)
```

File: optax/_src/transform.py [hunk 2]
```diff
@@ -364,6 +366,44 @@ def update_fn(updates, state, params=None):
   return GradientTransformation(init_fn, update_fn)
 
 
+def scale_by_belief(
+    b1: float = 0.9, b2: float = 0.999,
+    eps: float = 1e-8, eps_root: float = 0.0) -> GradientTransformation:
+  """Rescale updates according to the Adam algorithm.
+
+  References:
+    [Kingma et al, 2014](https://arxiv.org/abs/1412.6980)
```
Comment (mtthss) [inline optax/_src/transform.py:375]: Please link the Adabelief paper instead here

**Comment by mtthss** [comment]
Could you also expose the new function in the top level `__init__.py`?
(importing it there, and adding it to the public API defined by `__all__`).

**Comment by mtthss** [comment]
For testing the `adabelief` optimiser could you add it to the list of optimisers in `test_parabel` and `test_rosenbrock`?

File: optax/_src/transform.py [hunk 3]
```diff
@@ -393,10 +393,12 @@ def init_fn(params):
   def update_fn(updates, state, params=None):
     del params
     mu = _update_moment(updates, state.mu, b1, 1)
-    nu = _update_moment(updates - state.mu, state.nu, b2, 2)
+    prediction_error = jax.tree_multimap(lambda g, m: g-m, updates, state.mu)
+    nu = _update_moment(prediction_error, state.nu, b2, 2)
     count_inc = _safe_int32_increment(state.count)
     mu_hat = _bias_correction(mu, b1, count_inc)
-    nu_hat = _bias_correction(nu + eps, b2, count_inc)
+    nu_nonzero = jax.tree_map(lambda v: v + eps, nu)
```
Comment (mtthss) [inline optax/_src/transform.py:400]: Why are you adding an epsilon here as well?
The eps is already add in line 403.

Adam typically only adds it *after* the bias correction.
Why does adabelief need an additional epsilon term *before* the bias correction?

**Comment by mtthss** [comment]
Thanks! Great work!
One (final) comment

File: optax/_src/transform.py (hunk 3 above)
Comment (joaogui1) [reply optax/_src/transform.py:400]: There's this formula in the appendix:
image
But I believe that's just the `eps_root`, I hadn't noticed it existed before, my bad!

**Comment by joaogui1** [comment]
Thank you very much for the guidance and opportunity :)

**Comment by mtthss** [comment]
Amazing work, thanks for your contribution!
